### PR TITLE
Introduce block header V4 format to simplify block body verification.

### PR DESCRIPTION
**Pull request by robin-near**
The new block header V4 format includes the block body's hash into the block header. Today, all the block body's fields are already checked against corresponding fields in the block header (chunks against chunk_headers_root, challenges against challenges_root, and the vrf fields against random_value), so the new hash isn't actually necessary; however, it provides us with an easier and quicker way to verify the validity of the block body. We also now ensure that any further addition to the block body will be automatically covered by checks, which is nice given that this will be almost certain as we implement phase 2 of sharding.

Nayduck: https://nayduck.near.org/#/run/3131

### CODE REVIEW

**Comment by robin-near** [comment]
Latest run: https://nayduck.near.org/#/run/3132

File: pytest/tests/sanity/block_chunk_signature.py [hunk 1]
```diff
@@ -20,8 +21,9 @@ def __init__(self, *args, **kwargs):
         self.blocks = 0
 
     async def handle(self, msg, fr, to):
-        if msg.enum == 'Block':
-            msg.Block.BlockV2.chunks[0].signature.data = bytes(64)
+        if msg.enum == 'Block' and msg.Block.chunks(
+        )[0].signature.data != bytes(64):
+            msg.Block.chunks()[0].signature.data = bytes(64)
```
Comment (wacban) [inline pytest/tests/sanity/block_chunk_signature.py:26]: nit: funny formatting, maybe move msg.Block.chunks() to a helper variable

**Review by wacban — APPROVED** [review]
LGTM

**Comment by robin-near** [comment]
Latest nayduck build: https://nayduck.near.org/#/run/3142

I don't think the failures are related. If any of them are I'll fix afterwards.

**Pull request merged**